src/index.js

```diff
@@ -0,0 +1,126 @@
+function createCallback(onResult, extraProps) {
+  if (onResult) {
+    return (arg) => {
+      if (!extraProps.isCanceled) {
+        return onResult(arg);
+      }
+      return arg;
+    };
+  }
+}
+
+function thenFunc(extraProps, onSuccess, onError) {
+  return proxify(
+    this.then(
+      createCallback(onSuccess, extraProps),
+      createCallback(onError, extraProps)
+    ),
+    extraProps
+  );
+}
+
+function catchFunc(extraProps, onError) {
+  return proxify(this.catch(createCallback(onError, extraProps)), extraProps);
+}
+
+function finallyFunc(extraProps, onFinally) {
+  return proxify(
+    this.finally(createCallback(onFinally, extraProps)),
+    extraProps
+  );
+}
+
+function cancelFunc(extraProps) {
+  if (!extraProps.isCanceled) {
+    extraProps.isCanceled = true;
+    const controllerOpt = extraProps.options?.controller;
+    if (controllerOpt) {
+      // controller option can be a single controller or a list of controllers
+      const controllers = [].concat(controllerOpt);
+      for (const controller of controllers) {
+        controller?.abort();
+      }
+    }
+  }
+  return extraProps.isCanceled;
+}
+
+function proxify(promise, extraProps) {
+  /**
+   * A solution could be to use a Proxy but ES Proxy can't be polyfilled
+   * cf. from core-js (https://github.com/zloirock/core-js):
+   * > ES Proxy can't be polyfilled, you can try proxy-polyfill
+   * > which provides a very little subset of features.
+   *
+   * return new Proxy(promise, {
+   *   get(localPromise, prop) {
+   *     switch (prop) {
+   *       case 'then':
+   *         return thenFunc.bind(localPromise, extraProps);
+   *       case 'catch':
+   *         return catchFunc.bind(localPromise, extraProps);
+   *       case 'finally':
+   *         return finallyFunc.bind(localPromise, extraProps);
+   *       case 'cancel':
+   *         return cancelFunc.bind(localPromise, extraProps);
+   *       default:
+   *         return localPromise[prop];
+   *     }
+   *   },
+   * });
+   */
+  return {
+    then: thenFunc.bind(promise, extraProps),
+    catch: catchFunc.bind(promise, extraProps),
+    finally: finallyFunc.bind(promise, extraProps),
+    cancel: cancelFunc.bind(promise, extraProps),
+  };
+}
+
+export const makeCancelable = (promise, options) => {
+  const cancelablePromise = proxify(promise, { isCanceled: false, options });
+  const signalOpt = options?.signal;
+  if (signalOpt) {
```

Comment:
I don't think CancelablePromise should allows options to give some object to addEventListener. If you want to add a listener do it in your code.
First, `addEventListener` is a way to add Listener on DOM element. Maybe I use ReactiveX or some other libraries, and this is not useful for me.
Second, you never remove the event listener and there is no way to do it from my code. CancelablePromise seems to be a black box which add side effects. I don't think it's the idea of the library (add a `cancel()` method to promise).

For me, `makeCancelable` is useless, when you call it, you should call `proxify` that's all.

my 2 cents

Reply:
Yes I’m not convinced too, I’m just experimenting some stuff, I guess I will remove most of these features, and keep the lib as simple as possible. I prefer the functional approach over the class one, maybe  we should expose only 1 method, what do you think?

Reply:
It would be a big breaking change but we can expose `makeCancelable` and ignore class `CancelablePromise`.

```
makeCancelable(fetch(...));
```

Reply:
Ok I removed all the extra features and kept the current behavior, so no more breaking change, but I kept the functional approach too

```
import CancelablePromise from 'cancelable-promise';

new CancelablePromise((resolve, reject) => {});
new CancelablePromise(new Promise((resolve, reject) => {}));
CancelablePromise((resolve, reject) => {});
CancelablePromise(new Promise((resolve, reject) => {}));
```

Reply:
@immortal-tofu are you ok with this last iteration or do you have any other comments?

Reply:
Not fond of `CancelablePromise/new CancelablePromise`. It could be confusing to have same function which can be used with new or without new, with two types of arguments. Same for typings, it will be easier I think.
I'd prefer something like:
```javascript
const callback = (resolve, reject) => {};
cancelable(new Promise(callback));
new CancelablePromise(callback);
```

my 2 cents

Reply:
not fully agreed but I don't have a strong opinion about this, so ok I guess